autobahn/wamp_tcp_client.hpp

```diff
@@ -0,0 +1,111 @@
+    public:
+        wamp_tcp_client(std::shared_ptr<boost::asio::io_service> pIo,
+                        const boost::asio::ip::tcp::endpoint &Endpoint,
+                        const std::string &Realm,
+                        bool Debug = false) :
+                m_realm(Realm),
+                m_rawsocket_endpoint(Endpoint),
+                m_pIo(pIo),
+                m_debug(Debug) {
+            init();
+        }
+
+        wamp_tcp_client(std::shared_ptr<boost::asio::io_service> pIo,
+                        const std::string &IpAddress,
+                        uint16_t Port,
+                        const std::string &Realm,
+                        bool Debug = false) :
+                m_realm(Realm),
+                m_rawsocket_endpoint(boost::asio::ip::address::from_string(IpAddress), Port),
+                m_pIo(pIo),
+                m_debug(Debug) {
+            init();
+        }
+
+        /**
+     * @brief initialize members, creating shared pointers for the socket, and session
+     */
+        void init() {
+            m_pSession.reset();
+            m_pSocket.reset();
+            m_connected = boost::promise<bool>();
+            m_pSocket = std::make_shared<boost::asio::ip::tcp::socket>(*m_pIo);
+            m_pSession = std::make_shared<wamp_tcp_session_t>(*m_pIo, *m_pSocket, *m_pSocket, m_debug);
+        }
+
+        /**
+     * @brief launches the session asynchronously, returns a future containing an error code (0 means success)
+     */
+        boost::future<bool> launch() {
+            m_pSocket->async_connect(m_rawsocket_endpoint, [&](boost::system::error_code ec) {
+                if (!ec) {
+                    m_startFuture = m_pSession->start().then([&](boost::future<bool> started) {
+                        m_joinFuture = m_pSession->join(m_realm).then([&](boost::future<uint64_t> connected) {
+                            m_connected.set_value(true);
+                        });
+                    });
+                } else {
+                    std::cerr << "connect failed: " << ec.message() << std::endl;
+                    m_connected.set_value(false);
+                }
+            });
+            return m_connected.get_future();
+        }
+
+        ~wamp_tcp_client() {
+        }
+
+        /**
+     * This operator is used to pass through calls to the WAMP session. i.e *(pClient)->call(...)
+     */
+        std::shared_ptr<wamp_tcp_session_t> &operator->() {
+            return m_pSession;
+        }
+
+    private:
+        boost::future<void> m_startFuture; ///<holds the future of the start() operation
+        boost::future<void> m_joinFuture; ///<holds the future of the join() operation
+        boost::promise<bool> m_connected; ///<holds the future state of the success of launch
+        std::shared_ptr<wamp_tcp_session_t> m_pSession; //<need to be sure this is destructed before m_pSocket
```

Comment:
Not sure what the "p" is for in this and other member variables. I can only imagine that it would mean private? in which case it's unnecessary (and not consistent with other classes in AutobahnCpp) to include in the name, because all member variables are private anyway.